corehq/tabs/tabclasses.py

```diff
@@ -59,9 +64,42 @@ def sidebar_items(self):
             request=self._request, domain=self.domain)
         custom_reports = CustomProjectReportDispatcher.navigation_sections(
             request=self._request, domain=self.domain)
-        sidebar_items = tools + report_builder_nav + custom_reports + project_reports
+        sidebar_items = tools + report_builder_nav + self._regroup_sidebar_items(custom_reports + project_reports)
         return self._filter_sidebar_items(sidebar_items)
 
+    def _regroup_sidebar_items(self, sidebar_items):
+        try:
+            ordering = ReportsSidebarOrdering.objects.get(domain=self.domain)
+        except ObjectDoesNotExist:
+            return sidebar_items
+
+        reports_to_move = {}  # report class name to SidebarPosition mapping
+        for heading, reports in ordering.config:
+            for i, report_class_name in enumerate(reports):
+                reports_to_move[report_class_name] = SidebarPosition(heading, i)
+
+        # A mapping of headings to lists of 2-tuples. The tuples are index (under the heading) and the item itself.
+        new_sections = {x[0]: [] for x in ordering.config}
+        # A mapping of headings to lists of sidebar items.
+        old_sections = {x[0]: [] for x in sidebar_items}
+
+        # extract the sidebar items that to their new heading sections (or their old sections)
+        for heading, items in sidebar_items:
+            for item in items:
+                new_position = reports_to_move.get(item['class_name'], None)
+                if new_position:
+                    new_sections[new_position.heading].append((new_position.index, item))
+                else:
+                    old_sections[heading].append(item)
+
+        # Flatten the itermediary data structures
+        flat_new_sections = [
+            (heading, [item for index, item in sorted(new_sections[heading])])
+            for heading, _ in ordering.config
+        ]
+        flat_old_sections = [(heading, old_sections[heading]) for heading, _ in sidebar_items]
```

Comment:
Since its just getting the headings  from `sidebar_items`, `old_sections.keys()` and `x[0] for x in sidebar_items` are equivalent. Although probably more readable to just use `old_sections` throughout. I'll update